### PR TITLE
Fix: String#tr implementation

### PR DESCRIPTION
**Pull request by dnamsons**

### CODE REVIEW

File: artichoke-backend/src/extn/core/string/string.rb [hunk 1]
```diff
@@ -687,12 +687,11 @@ def to_str
   end
 
   def tr(from_str, to_str)
-    # TODO: Support character ranges c1-c2
```
Comment (dnamsons) [inline artichoke-backend/src/extn/core/string/string.rb:690]: It seems like character ranges have been implemented already, therefore i removed this TODO comment

File: spec-runner/vendor/spec/core/string/tr_spec.rb [hunk 2]
```diff
@@ -19,6 +19,7 @@
   it "pads to_str with its last char if it is shorter than from_string" do
     "this".tr("this", "x").should == "xxxx"
     "hello".tr("a-z", "A-H.").should == "HE..."
+    "abcd".tr("a-z", "xxx").should == "xxxx"
```
Comment (lopopolo) [inline spec-runner/vendor/spec/core/string/tr_spec.rb:22]: ruby-specs come from upstream. I'd consider this source file immutable. Can you back out this change?

If you think we're missing a test, add an integration test for `String` like these for `Array`:

https://github.com/artichoke/artichoke/blob/b99e3c1163381cec8ab505e33c2201dbe67c7ade/artichoke-backend/src/extn/core/array/mod.rs#L418-L440

https://github.com/artichoke/artichoke/blob/b99e3c1163381cec8ab505e33c2201dbe67c7ade/artichoke-backend/src/extn/core/array/array_functional_test.rb

File: artichoke-backend/src/extn/core/string/string.rb (hunk 1 above)
Comment (lopopolo) [reply artichoke-backend/src/extn/core/string/string.rb:690]: The below use of `Regexp` is a hack. This does method does not properly validate that ranges are ascending.

For example, this must error:

```
[2.6.6] > "hello".tr("a-y", "z-b")
Traceback (most recent call last):
        5: from /usr/local/var/rbenv/versions/2.6.6/bin/irb:23:in `<main>'
        4: from /usr/local/var/rbenv/versions/2.6.6/bin/irb:23:in `load'
        3: from /usr/local/var/rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):6
        1: from (irb):6:in `tr'
ArgumentError (invalid range "z-b" in string transliteration)
```

Can you restore the comment?

**Comment by lopopolo** [comment]
I tried this commit locally, and while it does improve spec compliance, there are still many ruby-spec failures. I'd like to take this PR with the changes I've requested.

Because `String#tr` still has some work to do, can you also edit the issue description to remove the "autoclose" annotation so GitHub does not close the attached ticket? Linking to it is fine.

**Comment by lopopolo** [comment]
Please squash and force push your changes so we don't add any extraneous commits to the specs directory.

**Comment by dnamsons** [comment]
@lopopolo Thanks for the review!

I've applied the changes you've asked

File: artichoke-backend/src/extn/core/string/string_test.rb [hunk 3]
```diff
@@ -42,4 +43,8 @@ def string_unary_minus
   raise unless s.itself == 'abababa'
 end
 
+def string_tr
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/string/string_test.rb:46]: oh I didn't realize we had one of these for `String` 🚀

**Comment by lopopolo** [comment]
link to https://github.com/artichoke/artichoke/issues/765

**Pull request merged**